### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Drew-Fisher**
Server backend overhaul to use dependency injection and ORM

### CODE REVIEW

File: chrome-extension/src/pages/content/components/constants.ts
Comment (JerecTharen) [inline chrome-extension/src/pages/content/components/constants.ts]: I purposely kept the constants file out of the repo. I don't really want this project associated with the company right off the bat. If we need this in the repo, then I would require it be changed this to be the profile url for LinkedIn's profile since that will help document that this is the only social media site that is supported.

File: server/src/app-datasource.ts
Comment (JerecTharen) [inline server/src/app-datasource.ts:3]: What's the different between this and the `ormconfig`? Could use a comment

File: server/src/controller/base-controller.ts
Comment (JerecTharen) [inline server/src/controller/base-controller.ts]: required *

File: server/src/controller/post-controller.ts
Comment (JerecTharen) [inline server/src/controller/post-controller.ts:44]: Could use the path to reach this endpoint in the comment here, should probably have that path in all the controller comments. Would help a lot. Took me a few minutes to figure all this out when I was testing it

I would also recommend adding any path or query parameters in the comment for `@req`

File: server/src/controller/post-controller.ts
Comment (JerecTharen) [inline server/src/controller/post-controller.ts]: paginated *

File: server/src/controller/post-controller.ts
Comment (JerecTharen) [inline server/src/controller/post-controller.ts:90]: Would recommend somehow documenting the format of the post object. Even if it is just adding the route to the post entity. Essentially, I want this controller file to be able to answer any questions at a glance that new developers have about how to use and maintain the back-end for these routes. Even if it does not answer all questions, it should at least be able to point them to where those answers can easily be found

File: server/src/controller/post-controller.ts
Comment (JerecTharen) [inline server/src/controller/post-controller.ts]: I think this should be the very first method in the class after the constructor. It is great documentation for how to use the routes and it's a bit hidden here at the bottom. Took me a second to figure out the paths and even after finding this once, I forgot about it when looking into another of the paths. Having it at the top would help that a lot. Or at least adding a comment at the top to go look at this method for documentation on the routes.

File: server/src/controller/post-controller.ts
Comment (JerecTharen) [inline server/src/controller/post-controller.ts]: The urn in the first route makes sense for it to be a route parameter, but when you have multiple parameters, I would highly recommend making them query parameters instead of path parameters.

File: server/src/dto/post-dto.ts
Comment (JerecTharen) [inline server/src/dto/post-dto.ts:8]: I've heard the term DTO a few times, and I am familiar with Entities from the .NET Entity Framework, but I think more documentation is needed here about the difference between the DTO and the Entity. From what I understand, the DTO is the object that would be referenced and manipulated in the back-end, while the Entity is the direct model used to represent the table in the database?

If so, say something along those lines here and in the Entity file

File: server/src/service/Ipost-service.ts
Comment (JerecTharen) [inline server/src/service/Ipost-service.ts:4]: POJO? I'm not familiar with that acronym.

File: server/src/repository/Ipost-repository.ts
Comment (JerecTharen) [inline server/src/repository/Ipost-repository.ts:4]: What is the difference between the post repository and the post service? Could probably use some documentation on that point.

**Review by JerecTharen — CHANGES_REQUESTED** [review]
I like what you put together here! A lot of good code that reminds me a ton of my .NET days 😄 

However, because this is a project meant for new developers to work on, I have added several notes about how to increase the documentation to assist them in the questions they may have, although in the end we will definitely need to still do a lot of pair programming with them.

In addition to that, I know you have yet to add README documentation for this, and before this get's merged I'd like for you to do so. Particularly I would like you to replace the README in the server directory with one that:

- Has a link back to the README for the original boilerplate
- Has setup instructions to assist with the docker containers
- Has instructions for how to navigate the back-end file structure to find the paths and the logic for them (just direct them to the controllers)
- Has instructions for how to add a new table to the database, or maintain the current one (it's okay here to heavily link to the documentation for type orm)

Additionally, I will be sending you a thunder client collection I put together for the posts. Thunder Client is a vs code extension/postman alternative that can used to store all the API Queries in a user friendly format. I would be obliged if you included the export in your MR and added instructions for how to import it and set up an environment (although I can do this last part after your PR goes through if you would like).

All in all though, this is great work and really gives us the start we need of the back-end! Thanks so much for putting it together!

File: server/src/service/Ipost-service.ts
Comment (Drew-Fisher) [reply server/src/service/Ipost-service.ts:4]: POJO means plane old Java/JavaScript object

File: server/src/repository/Ipost-repository.ts
Comment (Drew-Fisher) [reply server/src/repository/Ipost-repository.ts:4]: wrote a brief draft on this point let me know what you think

The service layer doesn't need to know how dat is presented or persisted which is why it is created independently of the persistence layers and presentation layers. Repositories define how data is persisted and retrieved, weather that be in an in memory array, database or local file. Separating these 2 concerns allows devs to easily switch out implementations and migrate the code base to different data sources for instance the only change to the service layer when migrating from a PostgreSQL data source to a NoSQL one would be to change the injected repository implementation.

File: server/src/controller/post-controller.ts
Comment (Drew-Fisher) [reply server/src/controller/post-controller.ts]: I made the adjustment and moved the routs method to the top of the class but now I think it might be better to include it in the BaseController class and override it in extending classes since it is required

File: server/src/app-datasource.ts
Comment (Drew-Fisher) [reply server/src/app-datasource.ts:3]: The difference is I actually got this one working lol. I am looking into the best option here, had some difficulties getting ormconfig working the way I wanted it to and used this as an alternative option.

File: chrome-extension/src/pages/content/components/constants.ts
Comment (Drew-Fisher) [reply chrome-extension/src/pages/content/components/constants.ts]: didn't realize this file was in the commit will remove it from the next one

File: server/src/controller/base-controller.ts
Comment (Drew-Fisher) [reply server/src/controller/base-controller.ts]: Corrected

File: server/src/controller/post-controller.ts
Comment (Drew-Fisher) [reply server/src/controller/post-controller.ts]: Corrected

File: server/src/controller/post-controller.ts
Comment (Drew-Fisher) [reply server/src/controller/post-controller.ts]: made the change to include the routes method in the BaseController class as an abstract method

File: server/src/controller/post-controller.ts
Comment (Drew-Fisher) [reply server/src/controller/post-controller.ts:44]: paths added to comments, migrated from path to query params

File: server/src/controller/post-controller.ts
Comment (Drew-Fisher) [reply server/src/controller/post-controller.ts]: migrated to query instead of path

File: server/src/app-datasource.ts
Comment (Drew-Fisher) [reply server/src/app-datasource.ts:3]: going to remove ormconfig.json for now may go back to this set up for typeorm later

**Comment by Drew-Fisher** [comment]
@JerecTharen changes have been made ready for another review

**Review by JerecTharen — APPROVED** [review]
Great, love the new README and especially the comments in the controller. I'm sure we'll need to make more changes moving forward, but this is a great first pull request into the main repo!

APPROVED 🎉

**Pull request merged**
